Add Copmpatibility for PDCurses

tysenmoore-xse opened the pull request:
I realize your code was written for NCurses however I was able to easily port your code to PDCurses which support Windows.  The changes were mainly in the build tools.  I was able to maintain all your NCurses support/demos.  If you are not interested I understand since this is titled "ncurses-st-menu".  I figured I'd give you the option.  Thanks for the time and most of all, thanks so much for the original implementation.

Commit:
- removed some files auto generated from autotools

Commit:
- Initial working port to PDCurses.  Ncurses works as well.

Commit:
- added initial version of Windows support
  - built with MinGW

Commit:
- Added back the demoapp_sl build for Windows
- Fixed the demoapp_sl for Linux

tysenmoore-xse committed:
Minor update to the README

tysenmoore-xse committed:
More README changes

okbob commented:
Hi, I have not problem merge your patch, it is great - just few notes:

1. I don't would to rename project :)
2. I hadn't possibility to fix a issues related to pd curses. So if I merge it, you should to fix some future bugs and issues.

okbob commented:
Hi, I reverted your pull request, because I have to fix simple demo first. After fixing, please, retry pull request.

GitMensch commented on demo/simple.c:
```diff
@@ -1,4 +1,8 @@
+
+#ifndef BUILD_OS_WIN
```
@tysenmoore-xse @okbob 
Please don't use a check for the build system here and in other places but check for the header and function in configure.ac, either directly or "more portable" (reference and hints for later: https://www.mail-archive.com/groff@gnu.org/msg01309.html)
... but the direct check for the header and function in configure.ac and the check of the auto-generated define would still be much better than checking the build system (which should also allow to get rid of those "self-created" defines [and possible breaking of unknown build systems that worked before] via configure at all)

okbob replied:
Unfortunately I have not good enough experience with autoconf and with writing portable code. I wrote and maintain only this library. So I know it, so my knowledges are too low. If you have any more specific ideas about enhancing or cleaning related issues, please, send pull request.

tysenmoore-xse replied:
@GitMensch
I did not consider this approach only because:
- the header was always used in the source so I assumed that was expected
- the header is not available in Windows
- this was only used in the demo/simple files, not core files
I can have a look at a configure.ac change to support this.  Seems like this is would not have been a unique comment to my PR but rather the original code.

GitMensch commented on include/st_panel.h:
```diff
@@ -0,0 +1,12 @@
+#elif defined HAVE_PANEL_H
+#include <panel.h>
+#else
+/* fallback */
```
@okbob @tysenmoore-xse 
If configure did not found the panel header then isn't it wrong to still include it?
Note: the old existing "error" in the original code from @okbob in st_menu.c was just changed:

```diff
 #if defined HAVE_NCURSESW_PANEL_H
 #include <ncursesw/panel.h>
 #elif defined HAVE_NCURSES_PANEL_H
 #include <ncurses/panel.h>
 #elif defined HAVE_PANEL_H
 #include <panel.h>
 #else
 /* fallback */
-#include <ncurses/panel.h>
+#include <panel.h>
 #endif
```

I suggest to place an `#error missing panel header` there, similar for st_curses.h.

okbob replied:
If I remember well, I did it when I worked on Solaris support. The correct solution should be fixing m4_ax_with_curses_extra.m4 script to detect this situation. But I am not able to do it :(. Importunately it is hell to work with ncurses headers. RH, Debian, FreeBSD, Solaris has different rules for placing of these files. Current state of pspg or ncurses-st-menu is little but messy, but it is almost work.

GitMensch replied:
Hm, you could still pass an option to configure: `CPPFLAGS="-DHAVE_NCURSES_PANEL_H` or similar in the case of "something does not work"; and if I did spot this right when I had a quick look in configure.ac: it would error and stop the building if it doesn't find the library/header.
@tysenmoore-xse I'd argue to at least don't change the "bad" fallback...

okbob commented:
I fixed some issues. Now, please, can you send your pull request again?

okbob commented:
but I checked your repository and ./autogen raise lot of warnings.
<pre>
configure.ac:19: warning: cannot check for file existence when cross compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
</pre>

okbob commented:
I tested your branch on my FC31 and there are some issues
<pre>
[pavel@nemesis ncurses-st-menu]$ make

Building: simple...
gcc demo/simple.c -o simple libst_menu.a  -Wall -L. -lpanelw -lncursesw -ltinfo  -L/usr/lib64 -lunistring -lXaw -lXmu -lXt -lX11 -lXpm -lSM -lICE -lXext -Iinclude -I. -g -O2   -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DPACKAGE_NAME=\"st_menu\" -DPACKAGE_TARNAME=\"st_menu\" -DPACKAGE_VERSION=\"0\" -DPACKAGE_STRING=\"st_menu\ 0\" -DPACKAGE_BUGREPORT=\"pavel.stehule@gmail.com\" -DPACKAGE_URL=\"\" -DSTDC_HEADERS=1 -DBUILD_OS_LINUX=1 -DXCURSES=1 -DHAVE_NCURSESW=1 -DHAVE_CURSES=1 -DHAVE_CURSES_ENHANCED=1 -DHAVE_CURSES_COLOR=1 -DHAVE_CURSES_OBSOLETE=1 -DHAVE_NCURSESW_CURSES_H=1 -DHAVE_CURSES_ENHANCED=1 -DHAVE_CURSES_COLOR=1 -DHAVE_CURSES_OBSOLETE=1 -DHAVE_NCURSES_H=1 -DHAVE_PANEL=1 -DHAVE_NCURSESW_PANEL_H=1 -DHAVE_ICONV=1 -DICONV_CONST= -DHAVE_LIBUNISTRING=1
/usr/bin/ld: cannot find -lXaw
/usr/bin/ld: cannot find -lXmu
/usr/bin/ld: cannot find -lXpm
collect2: error: ld returned 1 exit status
make: *** [Makefile:79: simple] Chyba 1
</pre>

tysenmoore-xse commented:
@okbob 
When you ran the make file and got the linker errors what were you building for?  NCurses? PDCurses?  I don't see the configure step so I am not sure what you are building.  I did not see these errors but can try to reproduce the error.  

As for the autogen warnings, I can have another look.  I did not run those scripts myself, I ran autoconf, etc. myself.

okbob commented:
út 25. 2. 2020 v 15:10 odesílatel Tysen Moore <notifications@github.com>
napsal:

> @okbob <https://github.com/okbob>
> When you ran the make file and got the linker errors what were you
> building for? NCurses? PDCurses? I don't see the configure step so I am not
> sure what you are building. I did not see these errors but can try to
> reproduce the error.
>

I try to work on fresh https://github.com/tysenmoore-xse/ncurses-st-menu.git

./configure
./make


 [pavel@nemesis ncurses-st-menu]$ ./configure
configure: loading site script /usr/share/config.site
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for ar... ar
checking the archiver (ar) interface... ar
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
configure: NCurses build...
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for ncursesw via pkg-config... yes
checking for working ncursesw/curses.h... yes
checking for working ncursesw.h... no
checking for working ncurses.h... yes
checking for Curses Panel library with ncursesw/panel.h... yes
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for shared library run path origin... done
checking for iconv... yes
checking for working iconv... yes
checking for iconv declaration...
         extern size_t iconv (iconv_t cd, char * *inbuf, size_t
*inbytesleft, char * *outbuf, size_t *outbytesleft);
checking for libunistring... yes
checking how to link with libunistring... /usr/lib64/libunistring.so
checking for libunistring version... 0.9.10
configure: creating ./config.status
config.status: creating config.make
config.status: creating st_menu.pc
[pavel@nemesis ncurses-st-menu]$ make

Building: simple...
gcc demo/simple.c -o simple libst_menu.a  -Wall -L. -lpanelw -lncursesw
-ltinfo  -L/usr/lib64 -lunistring -lXaw -lXmu -lXt -lX11 -lXpm -lSM -lICE
-lXext -Iinclude -I. -g -O2   -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600
 -DPACKAGE_NAME=\"st_menu\" -DPACKAGE_TARNAME=\"st_menu\"
-DPACKAGE_VERSION=\"0\" -DPACKAGE_STRING=\"st_menu\ 0\"
-DPACKAGE_BUGREPORT=\"pavel.stehule@gmail.com\" -DPACKAGE_URL=\"\"
-DSTDC_HEADERS=1 -DBUILD_OS_LINUX=1 -DXCURSES=1 -DHAVE_NCURSESW=1
-DHAVE_CURSES=1 -DHAVE_CURSES_ENHANCED=1 -DHAVE_CURSES_COLOR=1
-DHAVE_CURSES_OBSOLETE=1 -DHAVE_NCURSESW_CURSES_H=1
-DHAVE_CURSES_ENHANCED=1 -DHAVE_CURSES_COLOR=1 -DHAVE_CURSES_OBSOLETE=1
-DHAVE_NCURSES_H=1 -DHAVE_PANEL=1 -DHAVE_NCURSESW_PANEL_H=1 -DHAVE_ICONV=1
-DICONV_CONST= -DHAVE_LIBUNISTRING=1
/usr/bin/ld: cannot find -lXaw
/usr/bin/ld: cannot find -lXmu
/usr/bin/ld: cannot find -lXpm
collect2: error: ld returned 1 exit status
make: *** [Makefile:79: simple] Chyba 1

The problem is in config.make

HAVE_PDCURSES = no
PDCURSES_LIB =.
PDCURSES_DEP_LIBS = Xaw Xmu Xt X11 Xpm SM ICE Xext
PDCURSES_LIBDIR =.
PDCURSES_INCDIR =.

This looks little bit like nonsense - these libraries are necessary for
XWindows not for terminal application

so PDCURSES_DEP_LIBS = @PDCURSES_DEP_LIBS@ .. is probably wrong

but it starts few steps before

[pavel@nemesis ncurses-st-menu]$ ./autogen.sh
configure.ac:19: warning: cannot check for file existence when cross
compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross
compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross
compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross
compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross
compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level
configure.ac:19: warning: cannot check for file existence when cross
compiling
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/autoconf/general.m4:2777: AC_CHECK_FILE is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
tools/ax_pdcurses.m4:43: AX_PDCURSES is expanded from...
configure.ac:19: the top level





—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/okbob/ncurses-st-menu/pull/4?email_source=notifications&email_token=AAEFO45CVQDNZMNLKT4KIH3REURE3A5CNFSM4K2O6672YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEM4C7XY#issuecomment-590884831>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEFO43FSJPOZALSL6WAYBTREURE3ANCNFSM4K2O667Q>
> .
>

tysenmoore-xse commented:
Thanks for the additional information.  That was something missed in the ax_pdcurses.  That is a simple fix.  Thanks.

tysenmoore-xse commented:
I have addressed the following:

- Merged latest ncurses-st-menu changes
- Fixed warnings when running autogen.sh
- Fixed lib deps not getting cleared when not building PDCurses
- Removed use of BUILD_OS_WIN, now m4 check creating HAVE_LANGINFO_CODESET
- Added screenshot for X11 PDCurses
- Updates to README

Would it be worth another PR?  I can create another PR.  The only suggestion I did not address was the `#error missing panel header` suggestion since this seemed up in the air.

okbob commented:
út 25. 2. 2020 v 17:58 odesílatel Tysen Moore <notifications@github.com>
napsal:

> I have addressed the following:
>
>    - Merged latest ncurses-st-menu changes
>    - Fixed warnings when running autogen.sh
>    - Fixed lib deps not getting cleared when not building PDCurses
>    - Removed use of BUILD_OS_WIN, now m4 check creating
>    HAVE_LANGINFO_CODESET
>    - Added screenshot for X11 PDCurses
>    - Updates to README
>
> Would it be worth another PR? I can create another PR. The only suggestion
> I did not address was the #error missing panel header suggestion since
> this seemed up in the air.
>

It looks ok now - I tested you repository

please, create new PR

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/okbob/ncurses-st-menu/pull/4?email_source=notifications&email_token=AAEFO4YV2V3C2SMT24DFWTDREVE3FA5CNFSM4K2O6672YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEM4U7FI#issuecomment-590958485>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEFO4ZQUQGZBNRWWNLI4FDREVE3FANCNFSM4K2O667Q>
> .
>

tysenmoore-xse commented:
I'll create a new PR, thanks.  Sorry for the initial integration problems.

Also, would you like me to rework my README and maybe create a new section in the existing README for PDCurses?  I currently have my info at the front of the file.

tysenmoore-xse commented:
I created a new PR and did update the README.

okbob commented:
st 26. 2. 2020 v 3:20 odesílatel Tysen Moore <notifications@github.com>
napsal:

> I created a new PR and did update the README.
>
Please do rebase first, the merge was not possible.

And the code looks strange. Should be #ifdef there?

	#ifndef HAVE_LANGINFO_CODESET
	/* Don't use UTF when terminal doesn't use UTF */
	config.encoding = nl_langinfo(CODESET);
	#else
	config.encoding = "";
	#endif



> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/okbob/ncurses-st-menu/pull/4?email_source=notifications&email_token=AAEFO4ZJN2MISMXACSTAUOLREXGXRA5CNFSM4K2O6672YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEM6QXOA#issuecomment-591203256>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEFO46RIZR7WGDRO3PWSHDREXGXRANCNFSM4K2O667Q>
> .
>
